feat(modals): support stacking modals, remove bootstrap modals dependency

jaspervriends opened the pull request:
<!--
IMPORTANT: We applaud pull requests, they excite us every single time. As we have an obligation to maintain a healthy code standard and quality, we take sufficient time for reviews. Please do create a separate pull request per change/issue/feature; we will ask you to split bundled pull requests.
-->

Fixes #3516

**Changes proposed in this pull request:**
In this PR I've worked on making it possible to have multiple dialogs opened at the same time and make them stackable. I have partly used code of #3246 and altered it. It also removed the bootstrap modal dependency.

However, I did not proceed using the native HTML Dialog element for this PR due to lack of animations (for example, the backdrop is not animatable) and responsiveness with the interface.

A third parameter has been added to the `app.modal.show` function which is a boolean and defaults to `false`. This means that all already existing modals will keep their intended behaviour.

I also copied the new 'dismissable' options from the other PR, such as an option to hide the dialog on escape key, clicking on the backdrop or using the close button and managing them seperately per modal. This means there is a deprecation too which was mentioned in the code.

**Reviewers should focus on:**
- Current modals are opening and closing as intended
- Stacking multiple modals works
- Responsiveness is still good

**Necessity**

- [x] Has the problem that is being solved here been clearly explained?
- [x] If applicable, have various options for solving this problem been considered?
- [x] For core PRs, does this need to be in core, or could it be in an extension?
- [x] Are we willing to maintain this for years / potentially forever?

**Confirmed**

- [x] Frontend changes: tested on a local Flarum installation.
- [ ] Backend changes: tests are green (run `composer test`).
- [x] Core developer confirmed locally this works as intended.
- [x] Tests have been added, or are not appropriate here.

jaspervriends committed:
Feature: Stackable modals

davwheat commented on framework/core/js/src/common/components/ModalManager.tsx:
```diff
@@ -10,90 +10,154 @@ interface IModalManagerAttrs {
+      const Tag = modal?.componentClass;
+
+      // Make stackable modals
+      const zIndex = 9999 + modal.key;
```
Is this needed? Elements with the same zIndex should stack as per normal DOM hierarchy rules, where the last element is on top.

If we are going to do it this way, we should use the CSS variable for modal z-index and either:

1. Provide an custom attribute, such as `--modal-number`, then use `calc()` to set the z-index

2. Fetch the value of the modal z-index attribute on the `::root` selector and use that instead of a hardcoded value.

I expect number 1 will be easier.

davwheat commented on framework/core/js/src/common/components/ModalManager.tsx:
```diff
@@ -10,90 +10,154 @@ interface IModalManagerAttrs {
+      const zIndex = 9999 + modal.key;
+
+      return (
+        <div className="ModalManager modal" modal-key={modal.key} style={{ zIndex }}>
```
I would prefer to use `data-modal-key`.

davwheat replied:
We should also add `aria-hidden="true"` to all models except the active one.

Also `role="dialog"`.

davwheat commented on framework/core/js/src/common/components/ModalManager.tsx:
```diff
@@ -10,90 +10,154 @@ interface IModalManagerAttrs {
+  /**
+   * Get current active dialog
+   */
+  private get dialogElement(): HTMLElement {
```
Could we rename this to more suitable match the doc block?


```suggestion
  private get activeDialogElement(): HTMLElement {
```

davwheat commented on framework/core/js/src/common/components/ModalManager.tsx:
```diff
@@ -10,90 +10,154 @@ interface IModalManagerAttrs {
+  /**
+   * Get backdrop element of active dialog
+   */
+  private get backdropElement(): HTMLElement {
```
```suggestion
  private get activeBackdropElement(): HTMLElement {
```

davwheat commented on framework/core/js/src/common/states/ModalManagerState.ts:
```diff
@@ -19,44 +25,57 @@ export default class ModalManagerState {
+   * app.modal.show(MyCoolStackedModal, { attr: 'value' }, true);
    */
-  show(componentClass: UnsafeModalClass, attrs: Record<string, unknown> = {}): void {
+  show(componentClass: UnsafeModalClass, attrs: Record<string, unknown> = {}, stackModal: Boolean = false): void {
```
We should use `boolean` primitive rather than the `Boolean` object.

```suggestion
  show(componentClass: UnsafeModalClass, attrs: Record<string, unknown> = {}, stackModal: boolean = false): void {
```

davwheat reviewed (commented):
Just some comments so far. Haven't had time to properly review yet.

jaspervriends commented:
Thanks for the feedback, will work on it later this week

jaspervriends committed:
Processed feedback

jaspervriends commented:
Processed feedback :)

davwheat committed:
fix: use position in modal stack rather than key for modal number

davwheat committed:
fix: use correct base z-index

davwheat committed:
chore: simplify `className` to `class`

davwheat committed:
chore: add `key` attribute to ModalManager element

davwheat committed:
fix: backdrop flashing as modals are stacked/unstacked

davwheat committed:
chore: simplify modal close process

davwheat committed:
docs: add TS overload to indicate deprecated modal opening syntax

Require explicit values for `attrs` and `stackModal` from Flarum 2.0, beginning deprecation from now.

davwheat committed:
feat: use stackable modal for forgot password modal above sign in

davwheat reviewed (dismissed):
I've addressed some things with commits on this PR, but would like a sanity check for my changes from another core dev before looking to merge this.

Thank you for your work on this! This is a big improvement for Flarum's modals system! :)

davwheat committed:
chore: explicitly check if modal is open before trying to focus trap

davwheat committed:
fix(a11y): add missing  `aria-hidden` on main content when modal open

davwheat committed:
fix(a11y): add missing `aria-modal` on modal

SychO9 commented on framework/core/js/src/forum/components/LogInModal.tsx:
```diff
@@ -146,7 +146,7 @@ export default class LogInModal<CustomAttrs extends ILoginModalAttrs = ILoginMod
     const attrs = email.includes('@') ? { email } : undefined;
 
-    app.modal.show(ForgotPasswordModal, attrs);
+    app.modal.show(ForgotPasswordModal, attrs, true);
```
I don't believe this should be stackable 🤔

davwheat replied:
Sorry, was using this for testing! 🙈

SychO9 commented on framework/core/less/common/Modal.less:
```diff
@@ -6,55 +6,85 @@
   overflow: hidden;
```
`.modal-open` definition is dead code now I think

SychO9 commented on framework/core/js/src/common/states/ModalManagerState.ts:
```diff
@@ -19,62 +26,101 @@ export default class ModalManagerState {
+   * // This "hack" is needed due to quirks with nested redraws in Mithril.
+   * setTimeout(() => app.modal.show(MyCoolModal, { attr: 'value' }), 0);
+   */
+  show(componentClass: UnsafeModalClass, attrs?: Record<string, unknown>): void;
```
in 2.0 we'll probably keep the new `stackModal` argument set to `false` by default (or `true`). So I think we could dro this signature and the deprecation mark. Thoughts?

SychO9 reviewed (dismissed):
Code looks good! a few comments and a bug I noticed. Thanks a lot for the PR!

There's an issue with the scrolling behavior, I think a video will better explain it:
https://user-images.githubusercontent.com/20267363/178266180-69aaa829-b528-4808-999f-4693bd15b3a5.mp4

davwheat commented:
oops... that was the wrong git command...

have asked jasper for write access to his fork so i can resurrect this. have a backup of the code at https://github.com/flarum/framework/tree/jv/stackable-modals

davwheat committed:
chore: remove test code

davwheat committed:
chore: remove dead CSS

davwheat committed:
chore: remove overload

davwheat commented:
RE the scrolling bug: isn't this how the modals currently work in core?

https://user-images.githubusercontent.com/7406822/178990288-fede20f7-ccbc-4bf0-a72f-979446491969.mp4

SychO9 commented:
yea but that isn't how they work in this branch 🙈 `0:15` of my uploaded video shows behavior in this branch

davwheat commented:
OH I didn't watch for long enough! Right, I'll get on that.

davwheat committed:
fix: lock page scrolling when modal is open

davwheat commented:
@SychO9 Fixed, I hope.

SychO9 commented:
🙈 now you can't scroll inside a modal

davwheat committed:
fix: strange scrolling behaviour

davwheat committed:
chore: convert to JSX

davwheat committed:
fix: event listener memory leak

davwheat committed:
chore: remove unneeded optional chaining

davwheat committed:
fix: incorrect return types

davwheat committed:
chore: rewrite backdrop system - use one backdrop for all modals

davwheat committed:
docs: typos in comment block

SychO9 commented on framework/core/less/common/Modal.less:
```diff
@@ -45,41 +48,7 @@
-      }
-    }
-  }
+  pointer-events: none;
```
so the scrollbar shows up correctly but with `pointer-events: none` I can't scroll. But removing this also means you can't click outside to close 🤔

SychO9 commented on framework/core/js/src/common/components/ModalManager.tsx:
```diff
@@ -26,47 +26,62 @@ export default class ModalManager extends Component<IModalManagerAttrs> {
+            class="Modal-backdrop backdrop"
+            onclick={this.handlePossibleBackdropClick.bind(this)}
+            ontransitionend={this.onBackdropTransitionEnd.bind(this)}
+            data-showing={!!this.attrs.state.modalList.length}
```
Ideally, the backdrop would show behind the latest open modal and above the others. I think a simple CSS trick would do it: (other comment below for follow up)

```suggestion
            data-showing={!!this.attrs.state.modalList.length}
            style={{'--modal-count': this.attrs.state.modalList.length}}
```

SychO9 commented on framework/core/less/common/Modal.less:
```diff
@@ -14,24 +14,27 @@
+  opacity: 0;
+  transition: opacity 0.2s ease-out;
+  z-index: 1;
+  z-index: ~"calc(var(--zindex-modal) - 1)";
```
```suggestion
  z-index: ~"calc(var(--zindex-modal) + var(--modal-count)) - 2";
```

SychO9 reviewed (commented):
Using one backdrop seems great! I think only small tweaks are left.

davwheat committed:
fix: show backdrop directly below top-most modal

Co-authored-by: Sami Mazouz <sychocouldy@gmail.com>

davwheat committed:
chore: format

luceos commented:
The patience in this PR is admirable.

davwheat commented:
PS: JS checks fail in some cases due to attempting to commit and push dist JS to Jasper's repo since the PR is based on the `main` branch.

SychO9 committed:
fix: use an invisible backdrop for each modal to exit

Signed-off-by: Sami Mazouz <ilyasmazouz@gmail.com>

SychO9 committed:
chore: remove debugging code

Signed-off-by: Sami Mazouz <ilyasmazouz@gmail.com>

SychO9 commented:
Under known issues: https://caniuse.com/pointer-events
> Moving the scrollbar on an object with pointer-events: none; works in Firefox, but doesn't work in either Chrome or IE.

What I did is introduce invisible backdrops under each modal so that we don't have to use `pointer-events` and be able to easily exit the modal while maintaining scrollabilty (*is that a word?*).

davwheat commented:
Ahhhh that's a great find!

I didn't think of that idea and don't think I ever would have!

SychO9 commented on framework/core/js/src/common/components/ModalManager.tsx:
```diff
@@ -192,7 +196,7 @@ export default class ModalManager extends Component<IModalManagerAttrs> {
   }
 
-  protected handlePossibleBackdropClick(e: MouseEvent): void {
+  protected handlePossibleBackdropClick(e: MouseEvent | TouchEvent): void {
```
```suggestion
  protected handlePossibleBackdropClick(e: MouseEvent): void {
```

SychO9 committed:
chore: remove forgotten debug code

davwheat commented:
Rebased on `main`.

davwheat commented:
phew that was a big one 😄
